### PR TITLE
APEXMALHAR-2046 #resolve #comment adding spillable data structures fa…

### PR DESCRIPTION
**Pull request by ilooner**
…ctory interface.

### CODE REVIEW

**Comment by ilooner** [comment]
@chandnisingh please review.

File: library/src/main/java/org/apache/apex/malhar/lib/spillable/SpillableFactory.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  public void endWindow();
+
+  /**
+   * This is a factory method for creating a {@link SpillableArrayList}.
+   * @param <T> The type of data stored in the {@link SpillableArrayList}.
+   * @param bucket The bucket that this {@link SpillableArrayList} will be spilled too.
+   * @param serde The Serializer/Deserializer to use for data stored in the {@link SpillableArrayList}.
+   * @return A {@link SpillableArrayList}.
+   */
+  public <T> SpillableArrayList<T> newIndexedList(long bucket, Serde<T, byte[]> serde);
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/spillable/SpillableFactory.java]: If this doesn't need a prefix, will it create a default identifier?

File: library/src/main/java/org/apache/apex/malhar/lib/spillable/SpillableFactory.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+  /**
+   * This is a factory method for creating a {@link SpillableArrayList}.
+   * @param <T> The type of data stored in the {@link SpillableArrayList}.
+   * @param prefix The identifier for this {@link SpillableArrayList}.
+   * @param bucket The bucket that this {@link SpillableArrayList} will be spilled too.
+   * @param serde The Serializer/Deserializer to use for data stored in the {@link SpillableArrayList}.
+   * @return A {@link SpillableArrayList}.
+   */
+  public <T> SpillableArrayList<T> newIndexedList(byte[] prefix, long bucket, Serde<T, byte[]> serde);
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/spillable/SpillableFactory.java]: why `prefix`? why not `identifier`

File: library/src/main/java/org/apache/apex/malhar/lib/spillable/SpillableFactory.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  /**
+   * This is a factory method for creating a {@link SpillableByteIndexedMap}.
+   * @param <K> The type of the keys.
+   * @param <V> The type of the values.
+   * @param bucket The bucket that this {@link SpillableByteIndexedMap} will be spilled too.
+   * @param serdeKey The Serializer/Deserializer to use for the map's keys.
+   * @param serdeValue The Serializer/Deserializer to use for the map's values.
+   * @return A {@link SpillableByteIndexedMap}.
+   */
+  public <K, V> SpillableByteIndexedMap<K, V> newMap(long bucket, Serde<K, byte[]> serdeKey,
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/spillable/SpillableFactory.java]: what is a ByteIndexedMap?

File: library/src/main/java/org/apache/apex/malhar/lib/spillable/Serde.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * @param <SER> The type to Serialize an Object to.
+ */
+public interface Serde<OBJ, SER>
+{
+  /**
+   * Serialized the given object.
+   * @param object The object to serialize.
+   * @return The serialized representation of the object.
+   */
+  public SER serializeObject(OBJ object);
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/spillable/Serde.java]: Isn't `serialize` sufficient than `serializeObject`

File: library/src/main/java/org/apache/apex/malhar/lib/spillable/Serde.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  /**
+   * Deserializes the given serialized representation of an object.
+   * @param object The serialized representation of an object.
+   * @param offset An offset in the serialized reprsentation of the object.
+   * @return The deserialized object.
+   */
+  public OBJ deserializeObject(SER object, MutableInt offset);
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/spillable/Serde.java]: why MutableInt?

Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/spillable/Serde.java]: if this eases the following as @ilooner  mentioned 

```
Object obj;
MutableInt mi;
 someObj1 = deserialize(obj, mi);
 someObj2 = deserialize(obj, mi);
```

So the implementation is expected to modify the offset, then this needs to be documented.

File: library/src/main/java/org/apache/apex/malhar/lib/spillable/SpillableFactory.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+import com.google.common.collect.ListMultimap;
+
+import com.datatorrent.api.Component;
+import com.datatorrent.api.Context.OperatorContext;
+
+/**
+ * This is a factory for spillable data structures. This should be used as
+ * a component inside {@link com.datatorrent.api.Operator}s.
+ */
+public interface SpillableFactory extends Component<OperatorContext>
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/spillable/SpillableFactory.java]: IMO the spillable data-structures should be a component not the factory itself.

**Comment by ilooner** [comment]
@chandnisingh I have updated the interfaces with the suggestions from our discussions. Please let me know if I missed something, and merge if it looks good.

File: library/src/main/java/org/apache/apex/malhar/lib/spillable/SpillableComponentManager.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+import org.apache.apex.malhar.lib.spillable.Spillable.SpillableComponent;
+
+import com.datatorrent.api.Component;
+import com.datatorrent.api.Context.OperatorContext;
+
+/**
+ * This is a factory for spillable data structures. This should be used as
+ * a component inside {@link com.datatorrent.api.Operator}s.
+ */
+public interface SpillableComponentManager extends Component<OperatorContext>, SpillableComponent
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/spillable/SpillableComponentManager.java]: Where is SplillableComponent?

Documentation is outdated. It is not the factory. Is this the Spillable Composite Component ?  If yes, then can we call it a spillable composite component (or something to that effect) and not a spillable component manger

**Comment by ilooner** [comment]
@chandnisingh fixed mistakes.

**Comment by ilooner** [comment]
@chandnisingh fixed modifiers

**Pull request merged**